internal/login/auth.go

```diff
@@ -0,0 +1,141 @@
+package login
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	// "github.com/debricked/cli/internal/client"
+	"log"
+	"math/rand"
+	"net/http"
+	"os/exec"
+	"runtime"
+	"strings"
+	"time"
+
+	"golang.org/x/oauth2"
+)
+
+type IAuthenticator interface {
+	Authenticate() (*oauth2.Token, error)
+}
+
+type Authenticator struct {
+	ClientID string
+	Scopes   []string
+}
+
+const charset = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
+
+func generateRandomString(length int) string {
+	seed := rand.NewSource(time.Now().UnixNano())
+	r := rand.New(seed)
+
+	b := make([]byte, length)
+	for i := range b {
+		b[i] = charset[r.Intn(len(charset))]
+	}
+	return string(b)
+}
+
+func createCodeChallenge(codeVerifier string) string {
+	// Create a SHA-256 hash of the code verifier
+	hash := sha256.Sum256([]byte(codeVerifier))
+
+	// Encode the hash to base64
+	encoded := base64.StdEncoding.EncodeToString(hash[:])
+
+	// Make it URL safe
+	encoded = strings.TrimRight(encoded, "=")
+	encoded = strings.ReplaceAll(encoded, "+", "-")
+	encoded = strings.ReplaceAll(encoded, "/", "_")
+
+	return encoded
+}
+
+func (a Authenticator) Authenticate() (*oauth2.Token, error) {
+	// Set up OAuth2 configuration
+	config := &oauth2.Config{
+		ClientID:     a.ClientID,
+		ClientSecret: "",
+		Endpoint: oauth2.Endpoint{
+			AuthURL:  "https://debricked.com/app/oauth/authorize",
+			TokenURL: "https://debricked.com/app/oauth/token",
+		},
+		RedirectURL: "http://localhost:9096/callback",
+		Scopes:      a.Scopes,
+	}
+
+	// Create a random state
+	state := generateRandomString(8)
+	codeVerifier := generateRandomString(64)
```

Comment:
We could use the suggested `oauth2.GenerateVerifier()` here too: https://pkg.go.dev/golang.org/x/oauth2#GenerateVerifier. 

To make things simple, we could use it for the state too. It makes the state variable a bit longer, but I don't see that as anything negative. Then we can remove our `generateRandomString` function too :broom: 

```suggestion
	state := oauth2.GenerateVerifier()
	codeVerifier := oauth2.GenerateVerifier()
```